sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java

```diff
@@ -137,13 +139,79 @@ public AnalyzedRelation analyze(Node node, Analysis analysis) {
     }
 
     @Override
-    protected AnalyzedRelation visitQuery(Query node, StatementAnalysisContext context) {
-        return process(node.getQueryBody(), context);
+    protected AnalyzedRelation visitQuery(Query node, StatementAnalysisContext statementContext) {
+        // In case of Set Operation or VALUES clause
+        if (!node.getOrderBy().isEmpty() || node.getLimit().isPresent() || node.getOffset().isPresent()) {
+            QueriedRelation childRelation = (QueriedRelation) process(node.getQueryBody(), statementContext);
+
+            // Use child relation to process expressions of the "root" Query node
+            statementContext.startRelation();
+            RelationAnalysisContext relationAnalysisContext = statementContext.currentRelationContext();
+            String qualifiedName = childRelation.getQualifiedName().toString();
+            // In case of Union we need to analyze ORDER BY based on the output symbols of the left-most relation
+            if (childRelation instanceof UnionSelect) {
+                qualifiedName = ((UnionSelect) childRelation).left().getQualifiedName().toString();
+            }
+            relationAnalysisContext.addSourceRelation(qualifiedName, childRelation);
+            ExpressionAnalyzer expressionAnalyzer = new ExpressionAnalyzer(
+                functions,
+                statementContext.transactionContext(),
+                statementContext.convertParamFunction(),
+                new FullQualifiedNameFieldProvider(
+                    relationAnalysisContext.sources(),
+                    relationAnalysisContext.parentSources(),
+                    statementContext.transactionContext().sessionContext().defaultSchema()),
+                new SubqueryAnalyzer(this, statementContext));
+            ExpressionAnalysisContext expressionAnalysisContext = relationAnalysisContext.expressionAnalysisContext();
+            SelectAnalysis selectAnalysis = new SelectAnalysis(
+                childRelation.fields().size(),
+                relationAnalysisContext.sources(),
+                expressionAnalyzer,
+                expressionAnalysisContext);
+            for (Field field : childRelation.fields()) {
+                selectAnalysis.add(field.path(), field);
+            }
+            statementContext.endRelation();
```

Comment:
I think it would be better to call `endRelation` as early as possible (right after all sub-relations have been processed). It's easier to read if start/end calls stay close together.